### PR TITLE
upsert batch update to support array type

### PR DESCRIPTION
**Pull request by voyager-dw**
DataType DataValue UpsertBatch UpsertBatchBuilder change to support Array type for ingestion

### CODE REVIEW

File: metastore/common/data_type.go [hunk 1]
```diff
@@ -35,4 +35,19 @@ const (
 	GeoPoint  = "GeoPoint"
 	GeoShape  = "GeoShape"
 	Int64     = "Int64"
+
+	// array types
+	ArrayBool      = "ArrayBool"
```
Comment (jshencode) [inline metastore/common/data_type.go]: shall we do something like "Bool[]" or "Bool Array[]" instead?

File: memstore/common/data_type.go [hunk 2]
```diff
@@ -467,10 +542,53 @@ func GetGoDataValue(dataType DataType) GoDataValue {
...
+			if err != nil {
+				return nil, err
+			}
+			arrValue.AddItem(val)
+		}
+	}
+	return arrValue, nil
+}
+
+// ArrayValueFromString convert string to array of sepecified item data type
```
Comment (jshencode) [inline memstore/common/data_type.go]: add comment and example for the string form of array value

File: memstore/common/data_value.go [hunk 3]
```diff
@@ -132,6 +133,8 @@ type DataValue struct {
 	GoVal    GoDataValue
 	OtherVal unsafe.Pointer
 	DataType DataType
+	// only used for Array DataType, it is the data type of items inside array, the ItemType can not be Array and GeoShape for now
+	ItemType DataType
```
Comment (jshencode) [inline memstore/common/data_value.go]: we can get ItemType from DataType right?

File: memstore/common/data_value.go [hunk 4]
```diff
@@ -438,3 +480,210 @@ func (gs *GeoShapeGo) Write(dataWriter *utils.StreamDataWriter) error {
 	}
 	return dataWriter.WritePadding(int(dataWriter.GetBytesWritten()), 4)
 }
+
+// GetSerBytes return the bytes will be used in upsertbatch serialized format
+func (av *ArrayValue) GetSerBytes() int {
+	// there is a item number at beginning when serialize to upsert batch
+	return (4*8 + (DataTypeBits(av.DataType)*av.GetLength()+7)/8*8 + (av.GetLength()+7)/8*8 + 63) / 64 * 8
```
Comment (jshencode) [inline memstore/common/data_value.go]: please add more detailed comment for the calculation

File: memstore/common/data_value_test.go [hunk 5]
```diff
@@ -326,4 +326,153 @@ var _ = ginkgo.Describe("data value", func() {
...
+		s = CalculateListElementBytes(Int16, 10)
+		Ω(s).Should(Equal(24))
+
+		s = CalculateListElementBytes(Int32, 10)
+		Ω(s).Should(Equal(48))
+
+		s = CalculateListElementBytes(Int64, 10)
+		Ω(s).Should(Equal(88))
+
+		s = CalculateListElementBytes(Int64, 10)
```
Comment (shz117) [inline memstore/common/data_value_test.go]: deuplicated?

File: memstore/common/upsert_batch.go [hunk 6]
```diff
@@ -67,6 +67,21 @@ func (c *columnReader) ReadValue(row int) (unsafe.Pointer, bool) {
 	return unsafe.Pointer(&c.valueVector[row*DataTypeBits(c.dataType)/8]), true
 }
 
+// ReadArrayValue returns the ArrayValue from upsert batch at given row
+func (c *columnReader) ReadArrayValue(row int) (unsafe.Pointer, bool) {
+	validity := c.readValidity(row)
```
Comment (shz117) [inline memstore/common/upsert_batch.go]: is this validity of the array itself? I thought we don't have invalid arrays

File: memstore/common/upsert_batch_builder.go [hunk 7]
```diff
@@ -135,14 +141,14 @@ func (c *columnBuilder) CalculateBufferSize(offset *int) {
...
 func (c *columnBuilder) AppendToBuffer(writer *utils.BufferWriter) error {
 	writer.AlignBytes(1)
-	isGoType := IsGoType(c.dataType)
+	isVariableLength := IsGoType(c.dataType) || IsArrayType(c.dataType)
 
 	switch c.GetMode() {
 	case AllValuesDefault:
 		return nil
 	case HasNullVector:
 		// only non goType needs to write null vector
```
Comment (shz117) [inline memstore/common/upsert_batch_builder.go]: nit: update comment

File: memstore/common/upsert_batch_builder.go [hunk 8]
```diff
@@ -154,7 +160,7 @@ func (c *columnBuilder) AppendToBuffer(writer *utils.BufferWriter) error {
 	case AllValuesPresent:
 		var offsetWriter, valueWriter *utils.BufferWriter
 		// only goType needs to write offsetVector
```
Comment (shz117) [inline memstore/common/upsert_batch_builder.go]: nit: update comment

File: memstore/common/upsert_batch.go (hunk 6 above)
Comment (voyager-dw) [reply memstore/common/upsert_batch.go]: yeah, array should have validity too

File: memstore/common/upsert_batch_test.go [hunk 9]
```diff
@@ -608,4 +610,95 @@ var _ = ginkgo.Describe("upsert batch", func() {
...
+		value, err = upsertBatch.GetDataValue(0, 1)
+		Ω(err).Should(BeNil())
+		Ω(value).ShouldNot(BeNil())
+		Ω(value.Valid).Should(BeTrue())
+		reader := NewArrayValueReader(value.DataType, value.OtherVal)
+		data := make([]byte, 24)
+		for i := 0; i < 24; i++ {
+			data[i] = *(*byte)(unsafe.Pointer(uintptr(reader.value) + uintptr(i)))
+		}
+		fmt.Printf("data: %+v\n", data)
```
Comment (shz117) [inline memstore/common/upsert_batch_test.go]: nit

File: memstore/common/data_value.go [hunk 10]
```diff
@@ -438,3 +478,215 @@ func (gs *GeoShapeGo) Write(dataWriter *utils.StreamDataWriter) error {
...
+		DataType: dataType,
+		Items:    make([]interface{}, 0),
+	}
+}
+
+// Write serialize data into writer
+// Serialized Array data format:
+// number of items: 4 bytes
+// per item bytes * number of items
+// null bit * number of items
```
Comment (shz117) [inline memstore/common/data_value.go]: nit:
rename null/nill bit/array to `validity bit/array` to avoid confusion..

**Pull request merged**